File: benchmarks/java/PrimeBenchmark.java
import java.util.ArrayList;

public class PrimeBenchmark {
    public static void main(String[] args) {
        long startTime, endTime;

        // First run (not timed)
        countPrimes(1000000);

        // Second run (timed)
        startTime = System.nanoTime();
        countPrimes(1000000);
        endTime = System.nanoTime();

        System.out.printf("""
                {"type": "prime", "time_ns": %d}
                """, (endTime - startTime));
    }

    private static int countPrimes(int n) {
        boolean[] isPrime = new boolean[n + 1];
        for (int i = 2; i <= n; i++) {
            isPrime[i] = true;
        }
        for (int i = 2; i * i <= n; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j <= n; j += i) {
                    isPrime[j] = false;
                }
            }
        }
        int count = 0;
        for (int i = 2; i <= n; i++) {
            if (isPrime[i]) count++;
        }
        return count;
    }
}

File: benchmarks/scala/PrimeBenchmark.scala
import scala.util.Random

object PrimeBenchmark {
  def main(args: Array[String]): Unit = {
    val random = new Random()
    val randomMax = 10000000 + random.nextInt(101) // Random number between 10000000 and 10000100

    // First run (not timed)
    countPrimes(randomMax)

    // Second run (timed)
    val startTime = System.nanoTime()
    countPrimes(randomMax)
    val endTime = System.nanoTime()

    println(s"""{"type": "prime", "time_ns": ${endTime - startTime}}""")
  }

  def countPrimes(n: Int): Int = {
    val isPrime = Array.fill(n + 1)(true)
    for (i <- 2 to math.sqrt(n).toInt if isPrime(i)) {
      for (j <- i * i to n by i) {
        isPrime(j) = false
      }
    }
    isPrime.slice(2, n + 1).count(_ == true)
  }
}